File: examples/clean_api_python/infrastructure/seeders/product_seeder.py
"""Product seeder — Infrastructure / Seeders"""
from datetime import datetime
import sys, os

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../../..')))

from backbone.infrastructure.logging import LoggerFactory
from examples.clean_api_python.domain.entities.product import Product
from examples.clean_api_python.domain.repositories.product_repository import ProductRepository

_SEED_PRODUCTS = [
    {"id": "1", "name": "Laptop Dell XPS 15",    "description": "High performance laptop with 16GB RAM", "price": 1500.00, "category": "Electronics", "stock": 50},
    {"id": "2", "name": "iPhone 14 Pro",          "description": "Latest iPhone with advanced camera",    "price": 1200.00, "category": "Electronics", "stock": 100},
    {"id": "3", "name": "Office Chair Ergonomic", "description": "Comfortable ergonomic office chair",    "price": 350.00,  "category": "Furniture",   "stock": 30},
    {"id": "4", "name": "Standing Desk",          "description": "Adjustable height standing desk",       "price": 600.00,  "category": "Furniture",   "stock": 20},
    {"id": "5", "name": "Wireless Mouse",         "description": "Ergonomic wireless mouse",              "price": 45.00,   "category": "Electronics", "stock": 200},
]


class ProductSeeder:

    def __init__(self, repository: ProductRepository):
        self.repository = repository
        self.logger = LoggerFactory.create_layer_logger(
            "product-api", "infrastructure", "ProductSeeder"
        )

    def run(self) -> None:
        self.logger.info("Running product seeder...")
        now = datetime.now()
        seeded = 0

        for data in _SEED_PRODUCTS:
            try:
                product = Product(
                    id=data["id"],
                    name=data["name"],
                    description=data["description"],
                    price=data["price"],
                    category=data["category"],
                    stock=data["stock"],
                    created_at=now,
                    updated_at=now,
                )
                self.repository.create(product)
                seeded += 1
            except Exception as e:
                self.logger.warning("Failed to seed product", extra_data={
                    "product_id": data["id"], "error": str(e),
                })

        self.logger.info("Product seeder completed", extra_data={"seeded": seeded})

File: backbone-go/examples/clean-api-go/infrastructure/seeders/product_seeder.go
// Package seeders contains data seeders — Infrastructure layer.
package seeders

import (
	"context"
	"time"

	domainRepos "github.com/freakjazz/backbone-go/examples/clean-api-go/domain/repositories"
	"github.com/freakjazz/backbone-go/examples/clean-api-go/domain/entities"
	"github.com/freakjazz/backbone-go/infrastructure/logging"
)

var seedProducts = []struct {
	id, name, description, category string
	price                           float64
	stock                           int
}{
	{"1", "Laptop Dell XPS 15",    "High performance laptop with 16GB RAM", "Electronics", 1500.00, 50},
	{"2", "iPhone 14 Pro",          "Latest iPhone with advanced camera",    "Electronics", 1200.00, 100},
	{"3", "Office Chair Ergonomic", "Comfortable ergonomic office chair",    "Furniture",   350.00,  30},
	{"4", "Standing Desk",          "Adjustable height standing desk",       "Furniture",   600.00,  20},
	{"5", "Wireless Mouse",         "Ergonomic wireless mouse",              "Electronics", 45.00,   200},
}

// ProductSeeder seeds demo products into the repository.
type ProductSeeder struct {
	repository domainRepos.ProductRepository
	logger     *logging.EnhancedLogger
}

// NewProductSeeder creates a new seeder.
func NewProductSeeder(repo domainRepos.ProductRepository, logger *logging.EnhancedLogger) *ProductSeeder {
	return &ProductSeeder{repository: repo, logger: logger}
}

// Run seeds all demo products.
func (s *ProductSeeder) Run(ctx context.Context) {
	s.logger.Info("Running product seeder...", nil)
	now := time.Now()
	seeded := 0

	for _, d := range seedProducts {
		p := &entities.Product{
			ID:          d.id,
			Name:        d.name,
			Description: d.description,
			Price:       d.price,
			Category:    d.category,
			Stock:       d.stock,
			Active:      true,
			CreatedAt:   now,
			UpdatedAt:   now,
		}
		if err := s.repository.Create(ctx, p); err != nil {
			s.logger.Warning("Failed to seed product", map[string]interface{}{
				"id": d.id, "error": err.Error(),
			})
			continue
		}
		seeded++
	}

	s.logger.Info("Product seeder completed", map[string]interface{}{"seeded": seeded})
}
